examples/rust/providers/messaging-nats/src/nats.rs

```diff
@@ -0,0 +1,258 @@
+use core::time::Duration;
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::{anyhow, bail, Context as _};
+use async_nats::subject::ToSubject;
+use futures::StreamExt;
+use tokio::sync::RwLock;
+use tokio::task::JoinHandle;
+use tracing::{debug, error, warn};
+use wasmcloud::messaging::types::BrokerMessage;
+use wasmcloud_provider_sdk::core::HostData;
+use wasmcloud_provider_sdk::{
+    get_connection, load_host_data, run_provider, Context, LinkConfig, Provider,
+};
+
+use crate::connection::ConnectionConfig;
+
+wit_bindgen_wrpc::generate!();
+
+use exports::wasmcloud::messaging::consumer::Handler;
+
+/// [`NatsClientBundle`]s hold a NATS client and information (subscriptions)
+/// related to it.
+///
+/// This struct is necssary because subscriptions are *not* automatically removed on client drop,
+/// meaning that we must keep track of all subscriptions to close once the client is done
+#[derive(Debug)]
+struct NatsClientBundle {
+    pub client: async_nats::Client,
+    pub sub_handles: Vec<(String, JoinHandle<()>)>,
+}
+
+impl Drop for NatsClientBundle {
+    fn drop(&mut self) {
+        for handle in &self.sub_handles {
+            handle.1.abort()
+        }
+    }
+}
+
+/// Nats implementation for the `wasmcloud:messaging` interface
+#[derive(Default, Clone)]
+pub struct NatsMessagingProvider {
+    /// Map of NATS connection clients (including subscriptions) per component
+    components: Arc<RwLock<HashMap<String, NatsClientBundle>>>,
+    /// Default configuration to use when configuration is not provided on the link
+    default_config: ConnectionConfig,
+}
+
+impl NatsMessagingProvider {
+    /// Execute the provider, loading default configuration from the host and subscribing
+    /// on the proper RPC topics via `wrpc::serve`
+    pub async fn run() -> anyhow::Result<()> {
+        let host_data = load_host_data().context("failed to load host data")?;
+        let provider = Self::from_host_data(host_data);
+        let shutdown = run_provider(provider.clone(), "nats-messaging-provider")
+            .await
+            .context("failed to run provider")?;
+        let connection = get_connection();
+        serve(
+            &connection.get_wrpc_client(connection.provider_key()),
+            provider,
+            shutdown,
+        )
+        .await
+    }
+
+    /// Build a [`NatsMessagingProvider`] from [`HostData`]
+    pub fn from_host_data(host_data: &HostData) -> NatsMessagingProvider {
```

Comment:
Super nit (honestly more of a follow up for later): We have this function in most of our providers, but we don't have a function like `run_with_host_data` that does all the same things as `run` without fetching the host data